state/keys.go

```diff
@@ -56,40 +61,66 @@ func (k Keys) ChunkSizes() ([]uint16, bool) {
 	return chunks, true
 }
 
-type permsJSON []string
-
-type keysJSON struct {
-	Perms [8]permsJSON
-}
+type keysJSON map[string]Permissions
 
+// MarshalJSON marshals Keys has readable JSON.
+// Keys are hex encoded strings and permissions
+// are either valid named strings or unknown hex encoded strings.
 func (k Keys) MarshalJSON() ([]byte, error) {
-	var keysJSON keysJSON
+	kJSON := make(keysJSON)
 	for key, perm := range k {
-		keysJSON.Perms[perm] = append(keysJSON.Perms[perm], hex.EncodeToString([]byte(key)))
+		hexKey, err := codec.Bytes(key).MarshalText()
+		if err != nil {
+			return nil, err
+		}
+		kJSON[string(hexKey)] = perm
 	}
-	return json.Marshal(keysJSON)
+	return json.Marshal(kJSON)
 }
 
+// UnmarshalJSON unmarshals readable JSON.
 func (k *Keys) UnmarshalJSON(b []byte) error {
 	var keysJSON keysJSON
 	if err := json.Unmarshal(b, &keysJSON); err != nil {
 		return err
 	}
-	for perm, keyList := range keysJSON.Perms {
-		if perm < int(None) || perm > int(All) {
-			return fmt.Errorf("invalid permission encoded in json %d", perm)
+	for hexKey, perm := range keysJSON {
+		var key codec.Bytes
+		err := key.UnmarshalText([]byte(hexKey))
+		if err != nil {
+			return err
 		}
-		for _, encodedKey := range keyList {
-			key, err := hex.DecodeString(encodedKey)
-			if err != nil {
-				return err
-			}
-			(*k)[string(key)] = Permissions(perm)
+		(*k)[string(key)] = perm
+	}
+	return nil
+}
+
+func (p *Permissions) UnmarshalText(in []byte) error {
+	switch str := strings.ToLower(string(in)); str {
+	case "read":
+		*p = Read
+	case "write":
+		*p = Write
+	case "allocate":
+		*p = Allocate
+	case "all":
+		*p = All
+	case "none":
+		*p = None
+	default:
+		res, err := hex.DecodeString(str)
+		if err != nil || len(res) != 1 {
+			return fmt.Errorf("permission %s: %w", str, errInvalidHexadecimalString)
```

Comment:
Could we break this into two separate error cases so that we don't return an error that includes a nil error in the case where `len(res) != 1` ?

Reply:
I never include a `nil` error, we always wrap a `errInvalidHexadecimalString = errors.New("invalid hexadecimal string")` error